Reasoning Agent can execute code now.

BabyCNM opened the pull request:
<!-- Thank you for your contribution! Please review https://docs.ag2.ai/docs/contributor-guide/contributing before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've included any doc changes needed for https://docs.ag2.ai/. See https://docs.ag2.ai/docs/contributor-guide/documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

Hk669 commented:
@BabyCNM thanks for adding execution support to the reasoning agent, have you tested this locally?

BabyCNM commented:
@Hk669  Yes, I have tested locally. However, I want to include tool_use too. Hold on a bit.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ag2ai/ag2?pullRequest=545) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ag2ai/ag2?pullRequest=545) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ag2ai/ag2?pullRequest=545) it.</sub>

davorrunje commented:
@BabyCNM can you please sign the CLA: https://cla-assistant.io/ag2ai/ag2?pullRequest=545

sternakt commented on test/agentchat/contrib/test_reasoning_agent.py:
```diff
@@ -276,5 +276,58 @@ def test_visualize_tree_render_failure(mock_digraph):
         ])
 
 
+def test_reasoning_agent_code_execution():
+    """Test that ReasoningAgent properly executes code in responses"""
+    mock_config = {
+        "config_list": [{"model": "gpt-4o", "api_key": "fake", "base_url": "0.0.0.0:8000"}],
+        "temperature": 0,
+    }
```
To mock the credentials, use `mock_credentials: Credentials` fixture defined in [conftest.py](https://github.com/ag2ai/ag2/blob/main/test/conftest.py#L334)

[Here](https://github.com/ag2ai/ag2/blob/main/test/agents/experimental/deep_research/test_deep_research.py#L21) is an example of it being used in DeepResearchAgent test

marklysze commented:
@BabyCNM, a couple of things I'm experiencing...


1. The code execution example in the Notebook does execute code but the code produced didn't output anything, so I'm not sure if there should be additional prompting to make sure the answer is printed out.

E.g. this occurred a number of times:
```
...
mcts_agent (to reasoner_user_proxy):

Simulate the dice rolls with actual random values using Python and analyze the empirical results to estimate the expected maximum.
'''python
import random

def simulate_dice_rolls(num_rolls, num_trials):
    max_values = []
    for _ in range(num_trials):
        rolls = [random.randint(1, 6) for _ in range(num_rolls)]
        max_values.append(max(rolls))
    return sum(max_values) / len(max_values)

average_max = simulate_dice_rolls(3, 100000)
average_max
'''

--------------------------------------------------------------------------------

>>>>>>>> EXECUTING CODE BLOCK 0 (inferred language is python)...
reasoner_user_proxy (to mcts_agent):

exitcode: 0 (execution succeeded)
Code output: 


--------------------------------------------------------------------------------
```

2. I'm not sure if it just takes a very long time but the run just stopped after it was getting ready for next steps (?):
```
tot_thinker (to mcts_agent):

REFLECTION:
The previous steps evaluated the user's question regarding the expected maximum dice value when rolling a 6-sided die three times. The reflection indicated that the response was strong, providing correct calculations and probability breakdowns, but it hinted at areas for improvement in clarity and brevity. Although a fundamental understanding was established, the options could potentially enhance the depth and readability of the explanation.

**Possible Options:**
Option 1: Summarize the calculated probabilities for the maximum values in a clearer format to enhance readability and comprehension.

Option 2: Reiterate and validate the calculation process briefly to ensure understanding and accuracy before proceeding to detail the expected value.

Option 3: Correct the minor inconsistencies in detailing probabilities for \( k = 2 \) and \( k = 3 \), ensuring uniformity in the probability expressions.

Option 4: Provide an intuitive explanation or graphical illustration of the probability distribution of the maximum value to underpin the theoretical calculations.

Option 5: TERMINATE.

--------------------------------------------------------------------------------
mcts_agent (to tot_thinker):

## Here are some previous trajectories and reflections

### Previous Tries:
# Question:
What is the expected maximum dice value if you can roll a 6-sided dice three times?
---

Step 1: Reiterate the question regarding the expected maximum dice value and clarify your understanding of maximum in the context of independent rolls.
Step 2: TERMINATE.

Rating:Rating: 9/10

Explanation: 

The answer effectively addresses the original question about finding the expected maximum value of a 6-sided die rolled three times. It is factually accurate and complete, detailing the necessary steps to calculate the expectation with correct mathematical reasoning and logic.

**Strengths:**
- The step-by-step breakdown is clear and straightforward.
- The calculations for probabilities of each maximum value (from 1 to 6) are clearly laid out, using correct probability strategies related to independent events.
- The final expected value is computed correctly as approximately 4.25, which is accurate.

**Areas for Improvement:**
- While the calculations are largely correct, some of the probability expressions could be simplified for clarity, and the sections could be more concise to enhance readability and comprehension. 
- There is a minor inconsistency in detailing probabilities for \( k = 2 \) and \( k = 3 \) - the answer should maintain a uniform format for clarity.

Overall, it is a strong response but could benefit from a little more conciseness and clarity in the explanations for certain steps.



---

# Question:
What is the expected maximum dice value if you can roll a 6-sided dice three times?
---

Step 1: Calculate the expected maximum value by analyzing the distribution of outcomes when rolling the die three times.
Step 2: Summarize the calculated probabilities for the maximum values in a clearer format to enhance readability and comprehension.
---
What are the possible next steps?

--------------------------------------------------------------------------------
<Seem to hang here...>
```

BabyCNM commented:
@marklysze  Very good finding! Different models have different behavior in coding. I have updated the prompt and now it should work for most models. Regarding the timeout part, I did not have the issue. Maybe try to rerun to see if it is reproducible.

Hk669 reviewed (commented):
lets remove the cell outputs, i think its very hard to read and also navigate to different examples.

cc @BabyCNM

AgentGenie commented:
> lets remove the cell outputs, i think its very hard to read and also navigate to different examples.
> 
> cc @BabyCNM

I agree it is a bit hard to read.  But I think the cell outputs are very useful that users could figure out what the reasoning agent is doing without running the notebook themself.

AgentGenie commented:
I have a few questions,

1.  The last_meaningful_msg from the notebook seems to be convenient. Could it be provided from reasoning agent as the default method to use? i.e. as a static class method of reasoning agent.

2. I thought this PR is to enable tool use for reasoning agent but I don't understand what tool and where are those tool used?
I saw "F = TypeVar("F", bound=Callable[..., Any])" but doesn't see how it is used. Is it supposed to be the tool? Why not use ag2/autogen/tools/tool.py? I also don't see the tool use example from the notebook.

BabyCNM commented:
@AgentGenie The `last_meaningful_message` was originally part of the main agent but was later removed. We want to keep it configurable for different use cases rather than providing a hardcoded summary method.

BabyCNM commented:
@Hk669  @AgentGenie 
The rebase removed many comments, but I have addressed them now.
I removed the variable F and all outputs from the notebook.

@AgentGenie  The last_meaningful_message was originally part of the main agent but was later removed. We want to keep it configurable for different use cases rather than providing a hardcoded summary method.
